src/Content/Version.php

```diff
@@ -146,23 +165,88 @@ public function move(
 		$this->ensure($fromLanguage);
 		$this->model->storage()->move(
 			fromVersionId: $this->id,
-			fromLanguage: $this->language($fromLanguage),
+			fromLanguage: Language::ensure($fromLanguage),
 			toVersionId: $toVersionId,
-			toLanguage: $this->language($toLanguage)
+			toLanguage: Language::ensure($toLanguage)
 		);
 	}
 
 	/**
 	 * Returns the stored content fields
 	 *
 	 * @return array<string, string>
+	 */
+	public function read(Language|string $language = 'default'): array
+	{
+		$language = Language::ensure($language);
+
+		try {
+			$fields = $this->model->storage()->read($this->id, $language);
+			$fields = $this->convertFieldNamesToLowerCase($fields);
```

Comment:
`Thought`: TBH I'm not sure about converting to lowercase on read (on write is totally correct of course). At the moment, the `Data\Txt` class already handles normalization. And not just lowercasing but also normalization of dashes, underscore and spaces.

On the other hand, if we remove it here, we would run into issues with underlying storage handlers that don't normalize. Newly stored content should always be fine as it's also normalized on write, but existing content with different field name formatting would break merging of field arrays etc.

So maybe we do need this line here. I'm just not sure how far it should go, e.g. if we should also normalize the underscore/dash/space situation.

Reply:
So far, we do this in the Content class, which is a total mess and I never understand in which situations we actually convert the keys and when not. Putting it in the Version class felt like it's this single place where we take care of cleaning what's going into storage and what's coming out of storage and that felt nice. I have one more step on my list which is still missing here, but could maybe help with the decision. We also still have `ModelWithContent::contentFileData` which we also need to run here before storing content. Once that's implemented in the Version class as well, all the field normalization/sanitization/etc. is done here.

Reply:
Yesterday, while working on moving ::contentFieldData here, I thought that the method names could be better though. Maybe something like `::prepareFieldsToWrite` and `::prepareFieldsToRead`

Reply:
Good points. 👍

I fully agree it makes sense to have all the normalization in a single place. And `Version` is a great place for it.

I was thinking about the performance optimizations we currently have e.g. in `ModelWithContent::content()` to skip a second normalization because `Data\Txt::decode()` already does it. But maybe they are not needed as lowercasing array keys is probably not that intensive. And maybe PHP is even optimized to skip processing of array keys that are already lowercase.

Reply:
I think we really need to look into caching the version in memory after reading it anyway. Once we move `ModelWithContent::content` and `ModelWithContent::translation` to Version, we no longer have the $content property in the `ModelWithContent` class to take care of that. But that step is not yet ready.

Reply:
I think having a guaranteed normalization and reduced code complexity is worth more than a slight (probably not even noticeable) performance improvement.